### PR TITLE
enhance: cache image size and oci image spec in pouchd

### PR DESCRIPTION
**Pull request by fuweid**
Signed-off-by: Wei Fu <fuweid89@gmail.com>

### Ⅰ. Describe what this PR did

when the containerd manages many big images, the pouchd ListImage will
consume a lot of CPU resources caused by containerd.Content.Read stream
grpc call.

In the k8s, the kubelet gc image manager will send the ListImage request
every 30s. If the containerd consumes too many CPU resources, it will
impact the container. Therefore, pouchd needs to cache the data.

### Ⅱ. Does this pull request fix one issue?

NONE

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)

no need 

### Ⅳ. Describe how to verify it

CI

### CODE REVIEW

File: daemon/mgr/image_store.go [hunk 1]
```diff
@@ -62,6 +65,16 @@ type imageStore struct {
 
 	// primaryRefsIndexByID stores primay references, index by image ID
 	primaryRefsIndexByID map[digest.Digest]referenceMap
+
+	// cache ociImage to avoid open stream grpc to containerd, index by digest
```
Comment (allencloud) [inline daemon/mgr/image_store.go]: Please add the consequence which will happen if not do the cache thing.

File: daemon/mgr/image.go [hunk 2]
```diff
@@ -486,28 +493,22 @@ func (mgr *ImageManager) addReferenceIntoStore(id digest.Digest, ref reference.N
...
-	size, err := img.Size(ctx)
-	if err != nil {
-		return types.ImageInfo{}, err
-	}
-
-	ociImage, err := containerdImageToOciImage(ctx, img)
+func (mgr *ImageManager) containerdImageToImageInfo(ctx context.Context, id digest.Digest) (types.ImageInfo, error) {
+	ctrdImageInfo, err := mgr.localStore.GetCtrdImageInfo(id)
 	if err != nil {
+		if err == errCtrdImageInfoNotExist {
```
Comment (HusterWan) [inline daemon/mgr/image.go]: If the image not found in local store, should we check it from containerd?

File: daemon/mgr/image_store.go [hunk 3]
```diff
@@ -276,23 +293,43 @@ func (store *imageStore) RemoveReference(id digest.Digest, ref reference.Named)
 	return nil
 }
 
-// RemoveAllReferences removes all the reference by the given imageID.
-func (store *imageStore) RemoveAllReferences(id digest.Digest) error {
+// ListCtrdImageInfo returns all the CtrdImageInfo.
+func (store *imageStore) ListCtrdImageInfo() []CtrdImageInfo {
```
Comment (HusterWan) [inline daemon/mgr/image_store.go]: support filter later?

File: daemon/mgr/image.go [hunk 4]
```diff
@@ -211,6 +189,16 @@ func (mgr *ImageManager) RemoveImage(ctx context.Context, idOrRef string, force
 		return err
 	}
 
+	// since there is no rollback functionality, no guarantee that the
+	// containerd.RemoveImage must success. so if the localStore has been
+	// remove all the primary references, we should clear the CtrdImageInfo
+	// cache.
+	defer func() {
```
Comment (HusterWan) [inline daemon/mgr/image.go]: Whether we should remove the cache image info when the image has been succeeded?

File: daemon/mgr/image_store.go (hunk 3 above)
Comment (fuweid) [reply daemon/mgr/image_store.go]: YES!

**Pull request merged**